pkg/config/decode.go

```diff
@@ -0,0 +1,120 @@
+package config
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+
+	"github.com/pelletier/go-toml"
+)
+
+// decode parses a TOML tree (and environment variables) according to field tags set in the Config struct.
+// lookupEnv is used to look-up for env variables, if not provided it defaults to os.LookupEnv.
+// Essentially, decode expects `toml` and `env` field tags to be set. Cf. Config for reference.
+func decode(tree *toml.Tree, lookupEnv func(key string) (string, bool)) (conf Config, err error) {
+	// We recover from any panic during unmarshalling in order to return a proper error.
+	// Such cases are most likely due to a broken TOML configuration or invalid env variable value.
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
+		}
+	}()
+
+	if lookupEnv == nil {
+		lookupEnv = os.LookupEnv
+	}
+	d := &decoder{
+		tree:      tree,
+		lookupEnv: lookupEnv,
+	}
+
+	conf = DefaultConfig()
+	d.unmarshal(decodeContext{
+		val: reflect.ValueOf(&conf),
+		typ: reflect.TypeOf(&conf),
+	})
+
+	conf.tree = tree
+	return conf, nil
+}
+
+type decoder struct {
+	tree      *toml.Tree
+	lookupEnv func(key string) (string, bool)
+}
+
+type decodeContext struct {
+	val      reflect.Value
+	typ      reflect.Type
+	tomlPath []string
+	envVar   string
+}
+
+func (d *decoder) unmarshal(ctx decodeContext) {
```

Comment:
It seems weird that we have to do all of this work to unmarshal a toml document.

Reply:
At first I was considering to use third-party libs to unmarshal the config into a struct.
As we have 2 sources for the config (toml file and env variables to override),
it'd have required to unmarshal the struct twice, and the libraries I've found
had some quircks. For eg. https://github.com/caarlos0/env says :

> if the environment variable is not set, the zero-value of the type will be used

So it would overwrite the values set during the previous toml unmarshaling, unless we use
2 different structs and something like https://github.com/imdario/mergo. Eventually I felt
like it would be a cleaner solution to just have our own logic. The hard work is still being
done by the toml lib (parsing the configuration and generating a tree out of it), we just have
our own way of mapping that tree into a struct.